packages/xod-client-electron/arduino-libraries/ESP8266UART/ESP8266UART.h

```diff
@@ -0,0 +1,327 @@
+      uint8_t pos1 = 0;
+      uint8_t pos2 = 0;
+
+      // read chars until first match or timeout
+      uint32_t stop = millis() + timeout;
+      do {
+          while (_uart->available()) {
+              uint8_t c = 0x00;
+              bool res = _uart->readByte(&c);
+              debugChar((char) c);
+              if (!res) break;
+              pos1 = (c == buf1[pos1]) ? pos1 + 1 : 0;
+              pos2 = (c == buf2[pos2]) ? pos2 + 1 : 0;
+              if (len1 > 0 && pos1 == len1)
+                  return 1;
+              if (len2 > 0 && pos2 == len2)
+                  return 2;
+          }
+          // TODO: Replace with non-blicking delay?
+          delay(10);
+      } while (millis() < stop);
+      return 0;
+    }
+
+    uint8_t readBuffer(char* buff, uint8_t count, char delim, uint32_t timeout = 1000) {
+      uint8_t pos = 0;
+      uint8_t c;
+      uint32_t stop = millis() + timeout;
+      bool done = false;
+      do {
+        while (_uart->available() && pos < count) {
+          _uart->readByte(&c);
+          debugChar((char) c);
+          if (c == (uint8_t) delim) {
+            debug("B << ", "DELIMITER");
+            done = true;
+            break;
+          }
+          buff[pos++] = c;
+        }
+        if (done) break;
+      } while (millis() < stop);
+      buff[pos] = '\0';
+      return pos;
+    }
+
+  public:
+    ESP8266(xod::Uart& uart, Stream* debugSerial = NULL) {
+      _uart = &uart;
+      _debug = debugSerial;
+    }
+
+    void begin();
+    bool kick();
+
+    bool setStationMode();
+    bool connect(const char*, const char*);
+    bool isConnectedToAP();
+
+    uint32_t ipStringToUint32(char*);
+    char* getIPAddress();
+    uint32_t getIP();
+    char* getMAC();
+
+    bool createTCP(const char*, uint32_t);
+    bool releaseTCP();
+    bool send(char*);
+    bool read(char*);
+    bool isSocketOpen();
+    bool receive(char*);
+};
+
+void ESP8266::begin() {
+  _uart->begin();
+}
+
+bool ESP8266::kick() {
+  writeCmd(AT, EOL);
+  return cmdOK(OK, ERROR, 5000);
+}
+
+bool ESP8266::setStationMode() {
+  writeCmd(CWMODE_SET);
+  println("1");
+  return cmdOK(OK, ERROR, 1000);
+}
+
+bool ESP8266::connect(const char* ssid, const char* password) {
+  writeCmd(CWJAP);
+  print(ssid);
+  writeCmd(COMMA_2);
+  print(password);
+  writeCmd(DOUBLE_QUOTE, EOL);
+  cmdOK(OK, FAIL, 15000);
+  return isConnectedToAP();
+}
+
+
+bool ESP8266::isConnectedToAP() {
+  writeCmd(CIFSR, EOL);
+  uint8_t code = readCmd(NO_IP, ERROR, 350);
+  readCmd(OK, NULL, 10); // cleanup
+  return (code == 0);
+}
+
+uint32_t ESP8266::ipStringToUint32(char* ipAddress) {
+  unsigned int ip[4];
+  if ( 4 != sscanf(ipAddress, "%u.%u.%u.%u", &ip[0], &ip[1], &ip[2], &ip[3]) )
+       return 0;   // or some other indicator or error
+
+  if ( (ip[0] | ip[1] | ip[2] | ip[3]) >= 0x100 ) return 0;
+  return ip[0] + ip[1] * 0x100 + ip[2] * 0x10000ul + ip[3] * 0x1000000ul;
+}
+
+char* ESP8266::getIPAddress() {

```

Comment:
This method returns a pointer to the internal buffer which can mutate. A caller can cache it and later get an undesired behavior. To avoid it, make the method private as it indeed an implementation detail.